### ISSUE TITLE
Add a way to skip all email notifications

### ISSUE THREAD

**Issue by mulander** [plataformatec/devise#4526]
There is an existing API call to skip email confirmaton/reconfirmation notifications - that is a useful feature for administration panels where an admin might be making changes on behalf of the user but not wanting to generate email notifications.

The [skip_confirmation_notification!](http://www.rubydoc.info/github/plataformatec/devise/master/Devise/Models/Confirmable:skip_confirmation_notification!) works as advertised but doesn't prevent sending emails triggered by editing user passwords or emails. I don't want to disable those notifications site wide, just for a specific object in certain circumstances.

Currently I am working around this with the following addition to my user model:

```ruby
  @skip = false

  def skip_notifications!()
    skip_confirmation_notification!
    @skip = true
  end

  def email_changed?
    return false if @skip
    super
  end

  def encrypted_password_changed?
    return false if @skip
    super
  end
```

which allows me to disable all notifications on demand:

```
irb(main):002:0> u
#<User id: 20, email: "test@koparo.com", first_name: "Test", last_name: "User", created_at: "2017-05-03 17:37:14", updated_at: "2017-05-03 17:37:14", confirmed_age: true, accepted_tandc: true, role: "student">      
irb(main):004:0> u.email_changed?            
=> false                                     
irb(main):005:0> u.email = 'test2@koparo.com'
=> "test2@koparo.com"            
irb(main):006:0> u.email_changed?
=> true
irb(main):007:0> u.skip_notifications!
=> true
irb(main):008:0> u.email_changed?
=> false
irb(main):009:0> u.email = 'test3@koparo.com'
=> "test3@koparo.com"
irb(main):010:0> u.email_changed?
=> false
irb(main):011:0> exit

```

It would be nice if devise provided a gem wide API to silence all possible email notifications on demand in a single object scope.

Thanks for your continued work on this amazing gem.

**Comment by nashby** [comment]
Hey, looks like you need to call both `skip_confirmation!` and `skip_reconfirmation!` to achieve that.  Let me know if this doesn't help and I'll reopen this issue!

**Comment by mulander** [comment]
Hi @nashby,

I tested that before the initial report and just re-did the test again just to be sure by calling both `skip_confirmation!` and `skip_reconfirmation!` - with both devise still emails out password and email change notifications.

The method I called previously `skip_confirmation_notification!` is documented as:

> Skips sending the confirmation/reconfirmation notification email after_create/after_update. Unlike #skip_confirmation!, record still requires confirmation.

so it already internally disabled both, the note on 'still requires confirmation' means that the user will have to go through the confirmation process (which is nice and what I wanted) not that it needs a second call to disable it.

The emails I want to be able to disable are from the following initialization setup:

```ruby
  # Send a notification to the original email when the user's email is changed.
  config.send_email_changed_notification = true

  # Send a notification email when the user's password is changed.
  config.send_password_change_notification = true
```

Calling `skip_reconfirmation!` does not prevent those emails from being sent and I believe that a way to disable them on a per edit basis like with email confirmation/reconfirmation notifications would be a nice addition to the API.

**Comment by daniel-rikowski** [comment]
Agreed, it would be a nice addition.

In the meantime you can streamline your workaround by overriding `send_password_change_notification?`

Just add this to your model:

```ruby
def send_password_change_notification?
  return false if @skip
  super
end
```
(Same for `send_email_changed_notification?`)

**Comment by tegon** [comment]
This seems like a reasonable change. Is anyone willing to open a PR for this?

**Comment by iorme1** [comment]
Hi, I added a pull request for this issue. Let me know if I need to make any changes, thanks!

**Comment by tegon** [comment]
Closed via #4824
